test/filters/test_experimental_voxel_grid.cpp

```diff
@@ -0,0 +1,290 @@
+
+  // Copied from the old VoxelGrid as there is no dedicated method auto old_hash =
+  auto old_hash = [&](const PointXYZ& p) {
+    int ijk0 = static_cast<int>(std::floor(p.x * inverse_leaf_size[0]) -
+                                static_cast<float>(min_b[0]));
+    int ijk1 = static_cast<int>(std::floor(p.y * inverse_leaf_size[1]) -
+                                static_cast<float>(min_b[1]));
+    int ijk2 = static_cast<int>(std::floor(p.z * inverse_leaf_size[2]) -
+                                static_cast<float>(min_b[2]));
+
+    // Compute the centroid leaf index
+    return ijk0 * divb_mul[0] + ijk1 * divb_mul[1] + ijk2 * divb_mul[2];
+  };
+
+  for (size_t i = 0; i < cloud->size(); ++i) {
+    if (isXYZFinite(cloud->at(i))) {
+      const size_t new_hash = experimental::hashPoint(
+          cloud->at(i), inverse_leaf_size, min_b, divb_mul[1], divb_mul[2]);
+      EXPECT_EQ(new_hash, old_hash(cloud->at(i)));
+    }
+  }
+}
+
+TEST(LeafLayout, ExperimentalVoxelGridEquivalency)
+{
+  PointCloud<PointXYZ> new_out_cloud, old_out_cloud;
+
+  experimental::VoxelGrid<PointXYZ> new_grid;
+  pcl::VoxelGrid<PointXYZ> old_grid;
+  new_grid.setLeafSize(0.02f, 0.02f, 0.02f);
+  old_grid.setLeafSize(0.02f, 0.02f, 0.02f);
+  new_grid.setInputCloud(cloud);
+  old_grid.setInputCloud(cloud);
+  new_grid.setSaveLeafLayout(true);
+  old_grid.setSaveLeafLayout(true);
+  new_grid.filter(new_out_cloud);
+  old_grid.filter(old_out_cloud);
+
+  const auto new_leaf = new_grid.getLeafLayout();
+  const auto old_leaf = old_grid.getLeafLayout();
+  ASSERT_EQ(new_leaf.size(), old_leaf.size());
+
+  // Centroid indices are different from the old implememnt because of the order of
+  // iterating grids is different.
+  // But the index should still point to the same point
+
+  // leaf layout content
+  for (size_t i = 0; i < new_leaf.size(); ++i) {
+    if (old_leaf.at(i) == -1) {
+      EXPECT_EQ(new_leaf.at(i), -1);
+    }
+    else {
+      ASSERT_NE(new_leaf.at(i), -1);
+      const auto& new_pt = new_out_cloud.at(new_leaf.at(i));
+      const auto& old_pt = old_out_cloud.at(old_leaf.at(i));
+      EXPECT_POINT_EQ(new_pt, old_pt);
+    }
+  }
+
+  // getCentroidIndex
+  for (const auto& pt : *cloud) {
+    const auto& new_pt = new_out_cloud.at(new_grid.getCentroidIndex(pt));
+    const auto& old_pt = old_out_cloud.at(old_grid.getCentroidIndex(pt));
+    EXPECT_POINT_EQ(new_pt, old_pt);
+  }
+
+  for (const auto& pt : new_out_cloud) {
+    const Eigen::MatrixXi random_pt = Eigen::MatrixXi::Random(3, 1);
+
+    const auto new_idx1 = new_grid.getNeighborCentroidIndices(pt, random_pt)[0];
+    const auto new_idx2 =
+        new_grid.getNeighborCentroidIndices(pt.x, pt.y, pt.z, random_pt)[0];
+
+    const auto old_idx = old_grid.getNeighborCentroidIndices(pt, random_pt)[0];
+
+    EXPECT_EQ(new_idx1, old_idx);
+    EXPECT_EQ(new_idx2, old_idx);
+  }
+}
+
+// Update to std::tuple which can get by type in C++17
```

Comment:
http://en.cppreference.com/w/cpp/utility/tuple/get

I think C++14 already has that

Reply:
Done